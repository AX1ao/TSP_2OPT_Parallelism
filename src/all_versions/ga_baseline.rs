/// GA Version 1️⃣ with Elitism and Final 2-Opt
use rand::seq::SliceRandom;
use rand::Rng;
use rand::thread_rng;
use crate::all_versions::utils::*;

#[allow(dead_code)]
pub fn run_ga_baseline(
    cities: &[City],
    population_size: usize,
    generations: usize,
    base_mutation_rate: f64,
) -> Individual {
    let city_count = cities.len();
    let mut rng = thread_rng();

    let elitism_k = 5;
    let early_stop_patience = 100;
    let mut best_fitness_so_far = f64::MIN;
    let mut generations_without_improvement = 0;

    let mut population: Vec<Individual> = (0..population_size)
        .map(|_| {
            let mut tour: Vec<usize> = (0..city_count).collect();
            tour.shuffle(&mut rng);
            Individual::new(tour, cities)
        })
        .collect();

    for gen in 0..generations {
        let mut new_population = Vec::with_capacity(population_size);
        let mutation_rate = base_mutation_rate * (1.0 - gen as f64 / generations as f64);

        for _ in 0..population_size {
            let parent1 = tournament_selection(&population, 5);
            let parent2 = tournament_selection(&population, 5);

            let mut child_tour = order_crossover(&parent1.tour, &parent2.tour);

            if rng.gen::<f64>() < mutation_rate {
                swap_mutation(&mut child_tour);
            }

            let child = Individual::new(child_tour, cities);
            new_population.push(child);
        }

        // Elitism
        population.sort_by(|a, b| b.fitness.partial_cmp(&a.fitness).unwrap());
        let elites = population.iter().take(elitism_k).cloned();
        for (i, elite) in elites.enumerate() {
            new_population[i] = elite;
        }

        population = new_population;

        let best = population
            .iter()
            .max_by(|a, b| a.fitness.partial_cmp(&b.fitness).unwrap())
            .unwrap();

        if gen % 50 == 0 || gen == generations - 1 {
            println!(
                "Generation {}: Best distance = {:.4} | Mutation rate = {:.4}",
                gen,
                best.distance(),
                mutation_rate
            );
        }

        if best.fitness > best_fitness_so_far {
            best_fitness_so_far = best.fitness;
            generations_without_improvement = 0;
        } else {
            generations_without_improvement += 1;
        }

        if generations_without_improvement >= early_stop_patience {
            println!(
                "Early stopping at generation {} (no improvement in {} generations)",
                gen, early_stop_patience
            );
            break;
        }
    }

    // Final 2-opt refinement
    let best = population
        .into_iter()
        .max_by(|a, b| a.fitness.partial_cmp(&b.fitness).unwrap())
        .unwrap();

    let refined = two_opt(&best.tour, cities);
    Individual::new(refined, cities)
}
