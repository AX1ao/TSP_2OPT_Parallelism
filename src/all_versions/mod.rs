pub mod utils;
pub mod two_opt_seq;
pub mod two_opt_par_ver2;
pub mod optimized_multithread_2opt;
pub mod optimized_ver2_multi2opt;
pub mod random_insert_ver3_multi2opt;
pub mod par_prototype;
pub mod par_topk;
pub mod par_topkplus;
pub mod ga_baseline;
pub mod ga_config;
pub mod par_ga;
