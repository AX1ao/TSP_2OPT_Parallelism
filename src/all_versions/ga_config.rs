/// GA Version 2️⃣ Configurable with Logging + Delayed 2-Opt
use std::fs::File;
use std::io::{BufWriter, Write};
use rand::seq::SliceRandom;
use rand::Rng;
use rand::thread_rng;
use crate::all_versions::utils::*;

#[allow(dead_code)]
pub fn run_ga_config(
    cities: &[City],
    population_size: usize,
    generations: usize,
    base_mutation_rate: f64,
    elitism_k: usize,
    refine_start: usize,
    refine_every: usize,
    top_n: usize,
) -> Individual {
    let city_count = cities.len();
    let mut rng = thread_rng();

    let early_stop_patience = 100;
    let mut best_fitness_so_far = f64::MIN;
    let mut generations_without_improvement = 0;

    let mut population: Vec<Individual> = (0..population_size)
        .map(|_| {
            let mut tour: Vec<usize> = (0..city_count).collect();
            tour.shuffle(&mut rng);
            Individual::new(tour, cities)
        })
        .collect();

    for gen in 0..generations {
        let mutation_rate = base_mutation_rate * (1.0 - gen as f64 / generations as f64);
        let mut new_population = Vec::with_capacity(population_size);

        for _ in 0..population_size {
            let parent1 = tournament_selection(&population, 5);
            let parent2 = tournament_selection(&population, 5);
            let mut child_tour = order_crossover(&parent1.tour, &parent2.tour);

            if rng.gen::<f64>() < mutation_rate {
                swap_mutation(&mut child_tour);
            }

            let child = Individual::new(child_tour, cities);
            new_population.push(child);
        }

        // Elitism
        population.sort_by(|a, b| b.fitness.partial_cmp(&a.fitness).unwrap());
        for (i, elite) in population.iter().take(elitism_k).cloned().enumerate() {
            new_population[i] = elite;
        }

        population = new_population;

        // Apply 2-opt refinement after delay
        if gen >= refine_start && gen % refine_every == 0 {
            for i in 0..top_n.min(population.len()) {
                let refined = two_opt(&population[i].tour, cities);
                population[i] = Individual::new(refined, cities);
            }
        }

        // Status print every 50 generations or final
        let best = population
            .iter()
            .max_by(|a, b| a.fitness.partial_cmp(&b.fitness).unwrap())
            .unwrap();

        if gen % 50 == 0 || gen == generations - 1 {
            println!(
                "Generation {}: Best distance = {:.4} | Mutation rate = {:.4}",
                gen,
                best.distance(),
                mutation_rate
            );
        }

        // Early stopping
        if best.fitness > best_fitness_so_far {
            best_fitness_so_far = best.fitness;
            generations_without_improvement = 0;
        } else {
            generations_without_improvement += 1;
        }

        if generations_without_improvement >= early_stop_patience {
            println!(
                "Early stopping at generation {} (no improvement in {} generations)",
                gen, early_stop_patience
            );
            break;
        }
    }

    let best = population
        .into_iter()
        .max_by(|a, b| a.fitness.partial_cmp(&b.fitness).unwrap())
        .unwrap();

    let refined_tour = two_opt(&best.tour, cities);
    Individual::new(refined_tour, cities)
}