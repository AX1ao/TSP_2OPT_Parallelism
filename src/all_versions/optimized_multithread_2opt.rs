/// Multithreaded Version 2️⃣ with Lazy Swap Evaluation
use rayon::prelude::*;
use std::cmp::Ordering;
use rand::thread_rng;
use rand::seq::SliceRandom;
use rand::Rng;
use crate::all_versions::utils::*;

// this version to optimize the original code is try to avoid calculating the total distance
// after the swap op of each pair of edges
// the changes on the total cost after an operation to swap a pair of edges is the
// differences between the original pair of edge and current pair

// TODO: we can apply a lazy swap way that calculated all the possible pairs of swap that can
// make the result better, then choose them if the points do not overlap
// which means not exist in the previous selected range

// TODO: if there is a way to parallize the calculation of the total distance among the current shuffle

fn swap2edges(tour: &mut [usize], i: usize, j: usize) {
    tour[i..=j].reverse();
}
#[allow(dead_code)]
pub fn multi_2opt_optimized1(tour: &[usize], cities: &[City]) -> (Vec<usize>, f64) {
    let n = tour.len();
    let possibilities: Vec<(usize, usize)> = (1..n - 1)
        .into_par_iter()
        .flat_map_iter(|i| (i + 1..n).map(move |j| (i, j)))
        .collect(); 
    
    let repeat_times = 10;
    let sample_size = possibilities.len() / 2;

    let better_res = (0..32)
        .into_par_iter()
        .map(|_| {
            let mut rng = thread_rng();
            let mut tour: Vec<usize> = (0..n).collect();

            tour.shuffle(&mut rng); // keep these part the same

            let mut route = tour.clone(); // original shuffle for current thread
            let mut dis = compute_total_distance(&route, &cities); // ⬅️ updated here

            while true {
                let sampled = possibilities.choose_multiple(&mut rng, sample_size)
                    .cloned()
                    .collect::<Vec<_>>();

                let mut can_modify = vec![];

                for &(i, j) in &sampled {
                    if i == 0 || j + 1 >= n { continue; }

                    let non_swap = euclidean_distance(&cities[route[i - 1]], &cities[route[i]])
                        + euclidean_distance(&cities[route[j]], &cities[route[(j + 1) % n]]);
                    let after_swap = euclidean_distance(&cities[route[i - 1]], &cities[route[j]])
                        + euclidean_distance(&cities[route[i]], &cities[route[(j + 1) % n]]);

                    if non_swap - after_swap > 10.0 {
                        can_modify.push((i, j, non_swap - after_swap));
                    }
                }

                can_modify.sort_by(|a, b| b.2.partial_cmp(&a.2).unwrap());

                let mut used = vec![false; n];
                let mut selected = vec![];

                for &(i, j, _) in &can_modify {
                    if used[i - 1] || used[i] || used[j] || used[(j + 1) % n] {
                        continue;
                    }
                    selected.push((i, j));
                    used[i - 1] = true;
                    used[i] = true;
                    used[j] = true;
                    used[(j + 1) % n] = true;
                }

                for &(i, j) in &selected {
                    swap2edges(&mut route, i, j);
                }

                let new_dis = compute_total_distance(&route, &cities); // ⬅️ updated here
                if dis - new_dis < 1.0 { break; }
                dis = new_dis;
            }

            (route, dis)
        })
        .min_by(|a, b| a.1.partial_cmp(&b.1).unwrap())
        .unwrap();
    
    better_res
}
