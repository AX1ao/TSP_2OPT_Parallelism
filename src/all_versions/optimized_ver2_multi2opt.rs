/// Multithreaded Version 3️⃣
use rayon::prelude::*;
use std::cmp::Ordering;
use rand::thread_rng;
use rand::seq::SliceRandom;
use rand::Rng;
use crate::all_versions::utils::*;

// after doing some experiments on the optimize version1 of multithread 2opt
// when the number of cities grows over 1000, the result is pretty bad
// although the runtime improves a lot compared to the first version of multithread
// 2opt, the output result is even worse than the initial version
// the idea of operating multi-edges swap will not cause this sharply downgrade
// on the result, in fact, it should result in a lower cost compared to the initial version
// (although I did some experiments on the initial code to change the variable to get better results
// and did not apply it on the revised version)

// Thus, I think the key bottleneck of the version 1 is that we use fully shuffled initial route for all the threads
// the 2opt is a local optimized algorithm, so compared to the sequential version, the 
// multithread version can neither get performance gain nor better result

// what about construct some route that may already have some local optimization

fn swap2edges(tour: &mut [usize], i: usize, j: usize) {
    tour[i..=j].reverse();
}

fn get_initial_route(start_point: usize, cities: &[City]) -> Vec<usize> {
    // a randomized start_position to start a greedy algotrithm to get the initial route
    let n = cities.len();
    let mut initial_route = vec![];
    let mut vis = vec![false; n]; // to mark whether a city is visited
    initial_route.push(start_point);
    vis[start_point] = true;
    let mut prev = start_point;

    for _ in 1..n {
        let nxt = (0..n).filter(|&i| !vis[i]).min_by(|&a, &b| {
            euclidean_distance(&cities[prev], &cities[a])
                    .partial_cmp(&euclidean_distance(&cities[prev], &cities[b]))
                    .unwrap()
        }).unwrap(); // the idea is that get all the distance from unvisited cities to the previous added city
        // and choose the nearest city to push back to the vector 
        initial_route.push(nxt);
        vis[nxt] = true; // this city
        prev = nxt; // like the linked list, move the pointer to the next position as the previous city
    }
    initial_route
}
#[allow(dead_code)]
pub fn multi_2opt_optimized2_V2(tour: &[usize], cities: &[City]) -> (Vec<usize>, f64) {
    let n = tour.len();
    let possibilities: Vec<(usize, usize)> = (1..n - 1)
            .into_par_iter()
            .flat_map_iter(|i| (i + 1..n).map(move |j| (i, j)))
            .collect(); 
    
    let repeat_times = 10;
    let sample_size = possibilities.len() / 2;
    // the number of threads can be changed later, the original number of threads is 16
    let better_res = (0..32)
    .into_par_iter()
    .map(|_| {
        
        // we may construct a better initial route
        // and to ensure that all the threads are optimized themselves without
        // doing redundent works, we may also do some randomize work on the start position
        let mut rng = thread_rng();
        let mut initial_route = if rng.gen_bool(0.5) {
            let mut tmp = (0..n).collect::<Vec<_>>();
            tmp.shuffle(&mut rng);
            tmp
        } else {
            get_initial_route(rng.gen_range(0..n), &cities)
        };
        
        let mut route = initial_route.clone(); // original shuffle for current thread
        let mut dis = compute_total_distance(&route, &cities);
        while true { // continue using the repeated sampled based method
            let sampled = possibilities.choose_multiple(&mut rng, sample_size)
                            .cloned()
                            .collect::<Vec<_>>();

            let mut can_modify = vec![];

            for &(i, j) in &sampled {
                if i == 0 || j + 1 >= n { continue; }
               
                // in the previous version, if we compute the total_dis everytime we swap 2 edges is too costly. 
                // And since swap 2 edges only change the 2 distance among the total cost of the trip. 
                // we can try to compare the original dis of the 2 edge and after swap the 2 edges. 
                // but there is an issue, if we modify the original route immediately if the new route cost 
                // less than original version, then the next operation is not on the original version
                let non_swap = euclidean_distance(&cities[route[i - 1]], &cities[route[i]])
                        + euclidean_distance(&cities[route[j]], &cities[route[(j + 1) % n]]);
                let after_swap = euclidean_distance(&cities[route[i - 1]], &cities[route[j]])
                        + euclidean_distance(&cities[route[i]], &cities[route[(j + 1) % n]]);

                if non_swap - after_swap > 10.0 {
                    can_modify.push((i, j, non_swap - after_swap));
                }
            }

            can_modify.sort_by(|a, b| b.2.partial_cmp(&a.2).unwrap());

            let mut used = vec![false; n];
            let mut selected = vec![];

            for &(i, j, _) in &can_modify {
                if used[i - 1] || used[i] || used[j] || used[(j + 1) % n] {
                    continue;
                }
                selected.push((i, j));
                used[i - 1] = true;
                used[i] = true;
                used[j] = true; 
                used[(j + 1) % n] = true;
            }

            for &(i, j) in &selected {
                swap2edges(&mut route, i, j);
            }

            let new_dis = compute_total_distance(&route, &cities);
            if dis - new_dis < 1.0 { break; }
            dis = new_dis;
        }

        (route, dis)
    })
    .min_by(|a, b| a.1.partial_cmp(&b.1).unwrap())
    .unwrap();
    
    better_res
}

// res
//  half shuffle + half local optimized initial routes

// Generating 100 cities...
// Initial tour cost: 51275.63
// Sequential Version:
// Final tour cost: 8422.33
// Time taken: 872.18µs
// Optimized version 2 of multithread 2opt:
// Total cost: 8118.07
// Total duration: 16.70ms

// Generating 200 cities...
// Initial tour cost: 108291.95
// Sequential Version:
// Final tour cost: 11259.63
// Time taken: 3.45ms
// Optimized version 2 of multithread 2opt:
// Total cost: 11399.23
// Total duration: 28.55ms

// Generating 1000 cities...
// Initial tour cost: 514295.53
// Sequential Version:
// Final tour cost: 26271.03
// Time taken: 99.56ms
// Optimized version 2 of multithread 2opt:
// Total cost: 28491.26
// Total duration: 451.21ms


// Generating 10000 cities...
// Initial tour cost: 5169210.34
// Sequential Version:
// Final tour cost: 81551.06
// Time taken: 13.12s
// Optimized version 2 of multithread 2opt:
// Total cost: 87902.14
// Total duration: 41.35s


// Generating 20000 cities...
// Initial tour cost: 10401205.85
// Sequential Version:
// Final tour cost: 115590.88
// Time taken: 56.38s
// Optimized version 2 of multithread 2opt:
// Total cost: 124416.38
// Total duration: 183.75s