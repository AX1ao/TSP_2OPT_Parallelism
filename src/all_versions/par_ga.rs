/// GA Version 3️⃣ Fully Parallel GA with Delayed 2-Opt
use rayon::prelude::*;
use rand::seq::SliceRandom;
use rand::{Rng, thread_rng};
use crate::all_versions::utils::*;

#[allow(dead_code)]
pub fn run_ga_parallel(
    cities: &[City],
    population_size: usize,
    generations: usize,
    base_mutation_rate: f64,
    elitism_k: usize,
    refine_start: usize,
    refine_every: usize,
    top_n: usize,
) -> Individual {
    let city_count = cities.len();
    let mut rng = thread_rng();

    let early_stop_patience = 100;
    let mut best_fitness_so_far = f64::MIN;
    let mut generations_without_improvement = 0;

    let mut population: Vec<Individual> = (0..population_size)
        .map(|_| {
            let mut tour: Vec<usize> = (0..city_count).collect();
            tour.shuffle(&mut rng);
            Individual::new(tour, cities)
        })
        .collect();

    for gen in 0..generations {
        let mutation_rate = base_mutation_rate * (1.0 - gen as f64 / generations as f64);

        // Parallel child generation
        let new_population: Vec<Individual> = (0..population_size)
            .into_par_iter()
            .map(|_| {
                let parent1 = tournament_selection(&population, 5);
                let parent2 = tournament_selection(&population, 5);
                let mut child_tour = order_crossover(&parent1.tour, &parent2.tour);

                if rand::random::<f64>() < mutation_rate {
                    swap_mutation(&mut child_tour);
                }

                Individual::new(child_tour, cities)
            })
            .collect();

        // Elitism (sequential)
        population.sort_by(|a, b| b.fitness.partial_cmp(&a.fitness).unwrap());
        let mut new_population = new_population;
        for (i, elite) in population.iter().take(elitism_k).cloned().enumerate() {
            new_population[i] = elite;
        }

        population = new_population;

        // Parallel 2-opt refinement on top N individuals
        if gen >= refine_start && gen % refine_every == 0 {
            population
                .par_iter_mut()
                .take(top_n)
                .for_each(|ind| {
                    let refined = two_opt(&ind.tour, cities);
                    *ind = Individual::new(refined, cities);
                });
        }

        let best = population
            .iter()
            .max_by(|a, b| a.fitness.partial_cmp(&b.fitness).unwrap())
            .unwrap();

        if gen % 50 == 0 || gen == generations - 1 {
            println!(
                "Generation {}: Best distance = {:.4} | Mutation rate = {:.4}",
                gen,
                best.distance(),
                mutation_rate
            );
        }

        if best.fitness > best_fitness_so_far {
            best_fitness_so_far = best.fitness;
            generations_without_improvement = 0;
        } else {
            generations_without_improvement += 1;
        }

        if generations_without_improvement >= early_stop_patience {
            println!(
                "Early stopping at generation {} (no improvement in {} generations)",
                gen, early_stop_patience
            );
            break;
        }
    }

    let best = population
        .into_iter()
        .max_by(|a, b| a.fitness.partial_cmp(&b.fitness).unwrap())
        .unwrap();

    let refined = two_opt(&best.tour, cities);
    Individual::new(refined, cities)
}
