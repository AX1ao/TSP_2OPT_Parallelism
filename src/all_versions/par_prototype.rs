use crate::all_versions::utils::*;
use rayon::prelude::*;
use std::cmp::Ordering;

/// Basic 2-opt Version 1️⃣
/// Parallel 2-opt prototype: Applies the best swap per iteration.
/// Uses Rayon to evaluate all candidate (i, j) pairs concurrently.
#[allow(dead_code)]
pub fn par_prototype(tour: &[usize], cities: &[City]) -> (Vec<usize>, f64) {
    let mut improved = true;
    let mut new_tour = tour.to_vec();
    let mut loop_count = 0;
    const DELTA_THRESHOLD: f64 = 1e-6;

    while improved {
        loop_count += 1;
        if loop_count > 1000 {
            println!("⚠️ Breaking out to avoid infinite loop!");
            break;
        }

        improved = false;
        let n = new_tour.len();
        let candidates: Vec<(usize, usize)> = (1..n - 1)
            .flat_map(|i| (i + 1..n).map(move |j| (i, j)))
            .collect();

        if let Some((i_best, j_best)) = candidates
            .par_iter()
            .filter_map(|&(i, j)| {
                let a = new_tour[i - 1];
                let b = new_tour[i];
                let c = new_tour[j];
                let d = new_tour[(j + 1) % n];

                let current_dist = euclidean_distance(&cities[a], &cities[b])
                    + euclidean_distance(&cities[c], &cities[d]);
                let new_dist = euclidean_distance(&cities[a], &cities[c])
                    + euclidean_distance(&cities[b], &cities[d]);

                let delta = current_dist - new_dist;

                if delta > DELTA_THRESHOLD {
                    Some((delta, i, j))
                } else {
                    None
                }
            })
            .max_by(|(da, _, _), (db, _, _)| da.partial_cmp(db).unwrap_or(Ordering::Equal))
            .map(|(_, i, j)| (i, j))
        {
            new_tour[i_best..=j_best].reverse();
            improved = true;
        }
    }

    let final_cost = compute_total_distance(&new_tour, cities);
    (new_tour, final_cost)
}
