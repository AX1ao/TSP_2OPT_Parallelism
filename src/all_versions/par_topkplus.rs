use crate::all_versions::utils::*;
use rayon::prelude::*;
use std::cmp::Ordering;

/// Basic 2-opt Version 3️⃣
/// Top-K++ Smart Parallel 2-Opt with Early Cutoff and Deferred Application
/// 1. Filters candidate swaps by `delta_thresh`
/// 2. Selects top-k non-overlapping swaps
/// 3. Only commits changes if the full tour improves
#[allow(dead_code)]
pub fn par_topkplus(tour: &[usize], cities: &[City], k: usize, delta_thresh: f64) -> (Vec<usize>, f64) {
    let mut improved = true;
    let mut new_tour = tour.to_vec();
    let mut loop_count = 0;

    while improved {
        loop_count += 1;
        if loop_count > 1000 {
            println!("⚠️ Breaking out to avoid infinite loop!");
            break;
        }

        improved = false;
        let n = new_tour.len();

        // Step 1: Evaluate (i, j) pairs in parallel and filter by delta_thresh
        let mut candidates: Vec<(f64, usize, usize)> = (1..n - 1)
            .flat_map(|i| (i + 1..n).map(move |j| (i, j)))
            .collect::<Vec<_>>()
            .par_iter()
            .filter_map(|&(i, j)| {
                let a = new_tour[i - 1];
                let b = new_tour[i];
                let c = new_tour[j];
                let d = new_tour[(j + 1) % n];

                let current_dist = euclidean_distance(&cities[a], &cities[b])
                    + euclidean_distance(&cities[c], &cities[d]);
                let new_dist = euclidean_distance(&cities[a], &cities[c])
                    + euclidean_distance(&cities[b], &cities[d]);

                let delta = current_dist - new_dist;
                if delta > delta_thresh {
                    Some((delta, i, j))
                } else {
                    None
                }
            })
            .collect();

        if candidates.is_empty() {
            break;
        }

        // Step 2: Sort by delta (descending)
        candidates.sort_by(|a, b| b.0.partial_cmp(&a.0).unwrap_or(Ordering::Equal));

        // Step 3: Select top-k non-overlapping swaps
        let mut applied_flags = vec![false; n];
        let mut to_apply = vec![];
        let mut count = 0;

        for &(_, i, j) in &candidates {
            if count >= k {
                break;
            }
            if applied_flags[i..=j].iter().any(|&used| used) {
                continue;
            }
            to_apply.push((i, j));
            for idx in i..=j {
                applied_flags[idx] = true;
            }
            count += 1;
        }

        // Step 4: Simulate applying swaps and only commit if overall tour improves
        let mut simulated_tour = new_tour.clone();
        for (i, j) in &to_apply {
            simulated_tour[*i..=*j].reverse();
        }

        let new_cost = compute_total_distance(&simulated_tour, cities);
        let old_cost = compute_total_distance(&new_tour, cities);

        if new_cost < old_cost {
            new_tour = simulated_tour;
            improved = true;
        }
    }

    let final_cost = compute_total_distance(&new_tour, cities);
    (new_tour, final_cost)
}
