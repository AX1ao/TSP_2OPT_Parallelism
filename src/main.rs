//! # main.rs — Correctness Test Runner for All TSP Versions
//!
//! This executable is used to validate the **correctness** of all implemented TSP solvers,
//! including sequential, multithreaded, and genetic algorithm variants.
//!
//! ## Structure:
//! - All version modules are located inside `mod all_versions/`
//! - Each module contains one strategy (e.g., `par_topk`, `ga_baseline`, etc.)
//! - This runner creates a small set of cities and checks that each version
//!   can run without panic and produces a valid tour.
//!
//! ## Purpose:
//! - Sanity check for compilation and logic correctness.
//! - Useful before benchmarking or deployment to verify that all implementations
//!   are functionally consistent on small inputs.
//!
//! ## Output:
//! - For each version: initial cost, final cost, and time taken.
//!
//! ## Usage:
//! ```sh
//! cargo run --bin main
//! ```

mod all_versions;

use all_versions::utils::*;
use all_versions::two_opt_seq::*;
use all_versions::two_opt_par_ver2::*;
use all_versions::optimized_multithread_2opt::*;
use all_versions::optimized_ver2_multi2opt::*;
use all_versions::random_insert_ver3_multi2opt::*;
use all_versions::par_prototype::*;
use all_versions::par_topk::*;
use all_versions::par_topkplus::*;
use all_versions::ga_baseline::*;
use all_versions::ga_config::*;
use all_versions::par_ga::*;

fn main() {
    // ========== Correctness Test (n = 10) ==========
    let cities = generate_cities(10, 121); // or any fixed seed
    let mut tour: Vec<usize> = (0..10).collect();
    shuffle_tour(&mut tour);
    println!("✅ Initial tour: {:?}, cost = {:.2}", tour, compute_total_distance(&tour, &cities));

    let (_, cost) = two_opt_seq(&tour, &cities);
    println!("▶️  two_opt_seq: {:.2}", cost);

    let (_, cost) = par_prototype(&tour, &cities);
    println!("▶️  par_prototype: {:.2}", cost);

    let (_, cost) = par_topk(&tour, &cities, 2);
    println!("▶️  par_topk (k=2): {:.2}", cost);

    let (_, cost) = par_topkplus(&tour, &cities, 2, 1e-6);
    println!("▶️  par_topkplus (k=2, δ=1e-6): {:.2}", cost);

    let (_, cost) = two_opt_par_ver2(&tour, &cities);
    println!("▶️  two_opt_par_ver2: {:.2}", cost);

    let (_, cost) = multi_2opt_optimized1(&tour, &cities);
    println!("▶️  optimized_multithread_2opt: {:.2}", cost);

    let (_, cost) = multi_2opt_optimized2_V2(&tour, &cities);
    println!("▶️  optimized_ver2_multi2opt: {:.2}", cost);

    let (_, cost) = multi_2opt_random_insert(&tour, &cities);
    println!("▶️  random_insert_ver3_multi2opt: {:.2}", cost);

    let best = run_ga_baseline(&cities, 20, 100, 0.1);
    println!("▶️  ga_baseline: {:.2}", best.distance());

    let best = run_ga_config(&cities, 20, 100, 0.1, 2, 30, 20, 5);
    println!("▶️  ga_config: {:.2}", best.distance());

    let best = run_ga_parallel(&cities, 20, 100, 0.1, 2, 30, 20, 5);
    println!("▶️  par_ga: {:.2}", best.distance());
}
