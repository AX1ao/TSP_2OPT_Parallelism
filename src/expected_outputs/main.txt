✅ Initial tour: [5, 2, 9, 0, 4, 8, 1, 7, 3, 6], cost = 4165.56
▶️  two_opt_seq: 2286.57
▶️  par_prototype: 2286.57
⚠️ Breaking out to avoid infinite loop!
▶️  par_topk (k=2): 2941.48
▶️  par_topkplus (k=2, δ=1e-6): 2919.30
▶️  two_opt_par_ver2: 2286.57
▶️  optimized_multithread_2opt: 2286.57
▶️  optimized_ver2_multi2opt: 2286.57
▶️  random_insert_ver3_multi2opt: 2345.65
Generation 0: Best distance = 2804.4741 | Mutation rate = 0.1000
Generation 50: Best distance = 2583.7301 | Mutation rate = 0.0500
Generation 99: Best distance = 2583.7301 | Mutation rate = 0.0010
▶️  ga_baseline: 2583.73
Generation 0: Best distance = 3069.3564 | Mutation rate = 0.1000
Generation 50: Best distance = 2286.5679 | Mutation rate = 0.0500
Generation 99: Best distance = 2286.5679 | Mutation rate = 0.0010
▶️  ga_config: 2286.57
Generation 0: Best distance = 2939.2271 | Mutation rate = 0.1000
Generation 50: Best distance = 2286.5679 | Mutation rate = 0.0500
Generation 99: Best distance = 2286.5679 | Mutation rate = 0.0010
▶️  par_ga: 2286.57
